server/app/services/rpc/container_handler.rb

```diff
@@ -0,0 +1,126 @@
+require_relative 'fixnum_helper'
+require_relative 'container_info_mapper'
+
+module Rpc
+  class ContainerHandler
+    include Celluloid
+    include FixnumHelper
+    include Logging
+
+    def initialize(grid)
+      @grid = grid
+      @logs = []
+      @stats = []
+      @cached_container = nil
+      @db_session = ContainerLog.collection.session.with(
+        write: {
+          w: 0, fsync: false, j: false
+        }
+      )
+    end
+
+    # @param [Hash] data
+    def save(data)
+      info_mapper = ContainerInfoMapper.new(@grid)
+      info_mapper.from_agent(data)
+    end
+
+    # @param [Hash] data
+    def log(data)
+      container = cached_container(data['id'])
+      if container
+        if data['time']
+          created_at = Time.parse(data['time'])
+        else
+          created_at = Time.now.utc
+        end
+        @logs << {
+          grid_id: @grid.id,
+          host_node_id: container['host_node_id'],
+          grid_service_id: container['grid_service_id'],
+          container_id: container['_id'],
+          created_at: created_at,
+          name: container['name'],
+          type: data['type'],
+          data: data['data']
+        }
+        if @logs.size >= 5
+          flush_logs
+        end
+      end
+    end
+
+    def flush_logs
```

Comment:
It probably is, because I don't spot any celluloid task scheduling points in this Actor. I assume the `@db_session` operations will block the actor and prevent any concurrent `logs(...)` tasks from appending to the `@logs`.

It just needs to be really careful... insert some random `sleep` call somewhere in here, and you might end up with duplicated `ContainerLogs` entries in mongodb?